### PR TITLE
Light state sync and multi user op

### PR DESCRIPTION
**Pull request by r0ohafza**

### CODE REVIEW

File: packages/smart-vaults/src/library/MultiSignerSignatureLib.sol [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+            mask = (1 << signerIndex);
+            if (alreadySigned & mask != 0) return false;
+
+            if (MultiSignerLib.isValidSignature(_hash, $.signers[signerIndex], sigWrappers[i].signatureData)) {
+                alreadySigned |= mask;
+            } else {
+                return false;
+            }
+        }
+        return true;
```
[the remark this answers is not in the document]
Comment (r0ohafza) [reply packages/smart-vaults/src/library/MultiSignerSignatureLib.sol:76]: need to keep erc1271 verification and userOp separate, make sure we are not returning early for the latter case.

File: packages/smart-vaults/src/utils/LightSyncMultiSigner.sol [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+ * @title Multi Signer with light state sync functionality
+ * @author Splits
+ */
+abstract contract LightSyncMultiSigner is MultiSigner {
+    /* -------------------------------------------------------------------------- */
+    /*                                   STRUCTS                                  */
+    /* -------------------------------------------------------------------------- */
+
+    /// @notice Signer Update types
+    enum SignerUpdateType {
```
[the remark this answers is not in the document]
Comment (r0ohafza) [reply packages/smart-vaults/src/utils/LightSyncMultiSigner.sol:21]: yes

File: packages/smart-vaults/src/utils/LightSyncMultiSigner.sol [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+                _threshold: _threshold,
+                _hash: _getSignerUpdateHash(_signerUpdate, _nonce),
+                _signature: _signerUpdate.normalSignature
+            })
+        ) {
+            revert SignerSetUpdateValidationFailed(_signerUpdate);
+        }
+    }
+
+    function _processsSignerSetUpdateMemory(
```
[the remark this answers is not in the document]
Comment (r0ohafza) [reply packages/smart-vaults/src/utils/LightSyncMultiSigner.sol:137]: my spell checker was disabled 😢

File: packages/smart-vaults/src/vault/SmartVault.sol [hunk 4]
```diff
@@ -319,104 +342,134 @@ contract SmartVault is MultiSigner, RootOwner, ERC1271, UUPSUpgradeable, Receive
...
+    }
+
     /**
-     * @notice validates if the signature provided by the signer at `signerIndex` is valid for the hash.
+     * @notice validates if the given hash was signed by the signers.
      */
-    function _isValidSignature(
-        bytes32 _hash,
-        uint8 _signerIndex,
+    function _isValidSignature(bytes32 _hash, bytes calldata _signature) internal view override returns (bool) {
```
[the remark this answers is not in the document]
Comment (r0ohafza) [reply packages/smart-vaults/src/vault/SmartVault.sol:354]: ?

[the remark this answers is not in the document]
Comment (wminshew) [reply packages/smart-vaults/src/vault/SmartVault.sol:354]: lol sorry this was a note to myself to come back to look at this and i think i never did :(

File: packages/smart-vaults/src/utils/LightSyncMultiSigner.sol [hunk 5]
```diff
@@ -70,11 +70,12 @@ abstract contract LightSyncMultiSigner is MultiSigner {
 
     function _processSignerSetUpdates(SignerSetUpdate[] memory _signerUpdates) internal {
         MultiSignerLib.MultiSignerStorage storage $ = _getMultiSignerStorage();
-        uint256 noOfUpdates = _signerUpdates.length;
-        for (uint256 i; i < noOfUpdates; i++) {
+        uint256 numUpdates = _signerUpdates.length;
+        for (uint256 i; i < numUpdates; i++) {
             _validateSignerSetUpdate(_signerUpdates[i]);
             $.nonce += 1;
```
Comment (wminshew) [inline packages/smart-vaults/src/utils/LightSyncMultiSigner.sol:76]: should be able to do something like 
```
emit updateNonce(++$.nonce);
```
to save gas here i think?

**Pull request merged**